watertap/unit_models/ion_exchange_0D.py

```diff
@@ -544,594 +590,664 @@ def build(self):
+            def HTU(b, j):
+                return b.vel_bed / (
+                    pyunits.convert(
+                        b.resin_bulk_dens, to_units=pyunits.kg / pyunits.m**3
+                    )
+                    * b.rate_coeff[j]
+                )
+
+        if self.config.isotherm == IsothermType.freundlich:
+
+            @self.Expression(
+                self.target_ion_set, doc="Removed total mass of ion at resin exhaustion"
+            )  # Croll et al (2023), Eq.16
+            def mass_removed_total(b, j):
+                return (prop_in.flow_mass_phase_comp["Liq", j] / b.kinetic_param) * log(
+                    b.bed_capacity_param + 1
+                )
+
+            @self.Expression(
+                self.target_ion_set, doc="Freundlich base coeff estimation"
+            )
+            def freundlich_k(b, j):
+                mass_bed = pyunits.convert(
+                    b.bed_vol_tot * b.resin_bulk_dens, to_units=pyunits.kg
+                )
+                return b.mass_removed_total[j] / (
+                    mass_bed
+                    * prop_in.conc_mass_phase_comp["Liq", j] ** (1 / b.freundlich_n)
+                )
+
+        # ==========CONSTRAINTS==========
 
         @self.Constraint(
-            self.target_ion_set,
-            doc="Separation factor calc",
+            self.target_ion_set, doc="Mass transfer for regeneration stream"
         )
-        def eq_sep_factor(b, j):
-            return b.separation_factor[j] == 1 / b.langmuir[j]
+        def eq_mass_transfer_regen(b, j):
+            return (
+                regen.get_material_flow_terms("Liq", j)
+                == -b.process_flow.mass_transfer_term[0, "Liq", j]
+            )
 
         @self.Constraint(
-            self.target_ion_set,
-            doc="Langmuir isotherm",
+            doc="Isothermal assumption for regen stream",
         )
-        def eq_langmuir(b, j):
-            return b.separation_factor[j] * (
-                b.c_norm[j] * (1 - b.resin_eq_capacity / b.resin_max_capacity)
-            ) == (b.resin_eq_capacity / b.resin_max_capacity * (1 - b.c_norm[j]))
+        def eq_isothermal_regen_stream(b):
+            return prop_in.temperature == regen.temperature
 
-        @self.Constraint(doc="Pressure conservation")
-        def eq_press_conservation(b):
-            return b.properties_in[0].pressure == b.properties_out[0].pressure
+        @self.Constraint(
+            doc="Isobaric assumption for regen stream",
+        )
+        def eq_isobaric_regen_stream(b):
+            return prop_in.pressure == regen.pressure
 
-        @self.Constraint(doc="Temperature conservation")
-        def eq_temp_conservation(b):
-            return b.properties_in[0].temperature == b.properties_out[0].temperature
+        for j in inerts:
+            self.process_flow.mass_transfer_term[:, "Liq", j].fix(0)
+            self.regeneration_stream[0].get_material_flow_terms("Liq", j).fix(0)
```

Comment:
Also be aware that we may have to change this soon when MCAS updates to allow for mass or molar flow in material flow terms.

Reply:
Since we are fixing it to zero, wouldn't it not really matter for these lines? And/or couldn't it be changed via CONFIG? I guess I am not really familiar with the MCAS updates.

Reply:
Yea, in this specific case it doesn't matter, but it will for the adsorbed species .